temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientInvoker.java

```diff
@@ -356,20 +349,111 @@ public <R> UpdateOutput<R> update(UpdateInput<R> input) {
-    }
-    return new UpdateAsyncOutput<>(future);
+  public <R> PollWorkflowUpdateOutput<R> pollWorkflowUpdate(PollWorkflowUpdateInput<R> input) {
+    DataConverter dataConverterWithWorkflowContext =
+        clientOptions
+            .getDataConverter()
+            .withContext(
+                new WorkflowSerializationContext(
+                    clientOptions.getNamespace(), input.getWorkflowExecution().getWorkflowId()));
+
+    UpdateRef update =
+        UpdateRef.newBuilder()
+            .setWorkflowExecution(input.getWorkflowExecution())
+            .setUpdateId(input.getUpdateId())
+            .build();
+
+    WaitPolicy waitPolicy =
+        WaitPolicy.newBuilder()
+            .setLifecycleStage(
+                UpdateWorkflowExecutionLifecycleStage
+                    .UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED)
+            .build();
+
+    PollWorkflowExecutionUpdateRequest pollUpdateRequest =
+        PollWorkflowExecutionUpdateRequest.newBuilder()
+            .setNamespace(clientOptions.getNamespace())
+            .setIdentity(clientOptions.getIdentity())
+            .setUpdateRef(update)
+            .setWaitPolicy(waitPolicy)
+            .build();
+
+    CompletableFuture<PollWorkflowExecutionUpdateResponse> future = new CompletableFuture<>();
+
+    Deadline pollTimeoutDeadline = Deadline.after(input.getTimeout(), input.getTimeoutUnit());
+    pollWorkflowUpdateHelper(future, pollUpdateRequest, pollTimeoutDeadline);
+    return new PollWorkflowUpdateOutput(
+        future.thenApply(
+            (result) -> {
+              if (result.hasOutcome()) {
+                switch (result.getOutcome().getValueCase()) {
+                  case SUCCESS:
+                    Optional<Payloads> updateResult = Optional.of(result.getOutcome().getSuccess());
+                    R resultValue =
+                        convertResultPayloads(
+                            updateResult,
+                            input.getResultClass(),
+                            input.getResultType(),
+                            dataConverterWithWorkflowContext);
+                    return resultValue;
+                  case FAILURE:
+                    throw new WorkflowUpdateException(
+                        input.getWorkflowExecution(),
+                        input.getUpdateId(),
+                        input.getUpdateName(),
+                        dataConverterWithWorkflowContext.failureToException(
+                            result.getOutcome().getFailure()));
+                  default:
+                    throw new RuntimeException(
+                        "Received unexpected outcome from poll update request: "
+                            + result.getOutcome().getValueCase());
+                }
+              }
+              //
+              throw new RuntimeException("Received no outcome from server");
+            }));
+  }
+
+  private void pollWorkflowUpdateHelper(
```

Comment:
Just want to note I reviewed this carefully and while it doesn't exactly match what waiting for history close event is doing (which is synchronous, passes timeout/unit to timeout exception, etc), I think what is here is good.

Reply:
yeah I spent most of the time of this PR on this method and the retry logic. I can include timeout/unit to the timeout exception if the team think it is useful to have.

Reply:
Nah, no need. I am good with this. And I checked deadline vs user given timeout, how it's implemented on Go, etc and all looks good.